### PR TITLE
Migrate ha-form to lit

### PR DESCRIPTION
**Pull request by bramkragten**
First try, needs some testing.

### CODE REVIEW

File: src/auth/ha-auth-flow.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ import {
   css,
 } from "lit-element";
 import "@material/mwc-button";
-import "../components/ha-form";
+import "../components/form/ha-form";
```
Comment (balloob) [inline src/auth/ha-auth-flow.ts]:
```suggestion
import "../components/ha-form/ha-form";
```

File: src/components/form/ha-form-boolean.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  public focus() {
+    if (this._input) {
+      this._input.focus();
+    }
+  }
+
+  protected render(): TemplateResult {
+    return html`
+      <paper-checkbox checked=${this.data} @change=${this._valueChanged}
+        >${this.label}</paper-checkbox
```
Comment (balloob) [inline src/components/form/ha-form-boolean.ts]: Can we put new lines after `>` and before `<`

File: src/components/form/ha-form-boolean.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+import "@polymer/paper-checkbox/paper-checkbox";
+// Not duplicate, is for typing
+// tslint:disable-next-line
+import { PaperCheckboxElement } from "@polymer/paper-checkbox/paper-checkbox";
+
+@customElement("ha-form-boolean")
+export class HaFormBoolean extends LitElement implements HaFormElement {
+  @property() public schema!: HaFormSchema;
+  @property() public data!: { [key: string]: any };
```
Comment (balloob) [inline src/components/form/ha-form-boolean.ts]: this is not true ? This is a boolean.

File: src/components/form/ha-form-float.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+import "@polymer/paper-input/paper-input";
+// Not duplicate, is for typing
+// tslint:disable-next-line
+import { PaperInputElement } from "@polymer/paper-input/paper-input";
+
+@customElement("ha-form-float")
+export class HaFormFloat extends LitElement implements HaFormElement {
+  @property() public schema!: HaFormSchema;
+  @property() public data!: { [key: string]: any };
```
Comment (balloob) [inline src/components/form/ha-form-float.ts]: This is not true ?

File: src/components/form/ha-form-float.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import { fireEvent } from "../../common/dom/fire_event";
+
+import "@polymer/paper-input/paper-input";
+// Not duplicate, is for typing
+// tslint:disable-next-line
+import { PaperInputElement } from "@polymer/paper-input/paper-input";
+
+@customElement("ha-form-float")
+export class HaFormFloat extends LitElement implements HaFormElement {
+  @property() public schema!: HaFormSchema;
```
Comment (balloob) [inline src/components/form/ha-form-float.ts]: I think that we can improve our typing if we have a schema for `HaFormFloatSchema` etc. `HaFormSchema = HaFormBooleanSchema | HaFormFloatSchema`

File: src/components/form/ha-form-integer.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            ></ha-paper-slider>
+          </div>
+        `
+      : html`
+          <paper-input
+            type="number"
+            .label=${this.label}
+            .value=${this.data}
+            .required=${this.schema.required}
+            .auto-validate=${this.schema.required}
```
Comment (balloob) [inline src/components/form/ha-form-integer.ts]: Polymer will convert kebab-case attributes to camelCase property names.
```suggestion
            .autoValidate=${this.schema.required}
```

File: src/components/form/ha-form-select.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+  public focus() {
+    if (this._input) {
+      this._input.focus();
+    }
+  }
+
+  protected render(): TemplateResult {
+    return html`
+      <paper-dropdown-menu label=${this.label}>
```
Comment (balloob) [inline src/components/form/ha-form-select.ts]:
```suggestion
      <paper-dropdown-menu .label=${this.label}>
```

File: src/components/form/ha-form-select.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        <paper-listbox
+          slot="dropdown-content"
+          attr-for-selected="item-name"
+          .selected=${this.data}
+          @change=${this._valueChanged}
+        >
+          ${this.schema.options!.map(
+            (item) => html`
+              <paper-item .item-name=${this._optionValue(item)}
+                >${this._optionLabel(item)}</paper-item
```
Comment (balloob) [inline src/components/form/ha-form-select.ts]: New line after `>` and before `<`

File: src/components/form/ha-form-select.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+      <paper-dropdown-menu label=${this.label}>
+        <paper-listbox
+          slot="dropdown-content"
+          attr-for-selected="item-name"
+          .selected=${this.data}
+          @change=${this._valueChanged}
+        >
+          ${this.schema.options!.map(
+            (item) => html`
+              <paper-item .item-name=${this._optionValue(item)}
```
Comment (balloob) [inline src/components/form/ha-form-select.ts]:
```suggestion
              <paper-item .itemName=${this._optionValue(item)}
```

File: src/components/form/ha-form-string.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+  protected render(): TemplateResult {
+    return this.schema.name.includes("password")
+      ? html`
+          <paper-input
+            .type=${this._unmaskedPassword ? "text" : "password"}
+            .label=${this.label}
+            .value=${this.data}
+            .required=${this.schema.required}
+            .auto-validate=${this.schema.required}
```
Comment (balloob) [inline src/components/form/ha-form-string.ts]:
```suggestion
            .autoValidate=${this.schema.required}
```

File: src/components/form/ha-form-string.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+            >
+            </paper-icon-button>
+          </paper-input>
+        `
+      : html`
+          <paper-input
+            .label=${this.label}
+            .value=${this.data}
+            .required=${this.schema.required}
+            .auto-validate=${this.schema.required}
```
Comment (balloob) [inline src/components/form/ha-form-string.ts]:
```suggestion
            .autoValidate=${this.schema.required}
```

File: src/components/form/ha-form.ts [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+      <div id="element" @value-changed=${this._valueChanged}></div>
+    `;
+  }
+
+  protected updated(changedProperties: PropertyValues) {
+    const oldSchema: HaFormSchema = changedProperties.get(
+      "schema"
+    ) as HaFormSchema;
+    if (
+      !Array.isArray(this.schema) &&
```
Comment (balloob) [inline src/components/form/ha-form.ts]: Guard clause it. `if (isArray) return`

Comment (balloob) [inline src/components/form/ha-form.ts]: What if the current schema is an array, shouldn't we clean up the element ?

File: src/components/form/ha-form.ts [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+      : schema
+      ? schema.name
+      : "";
+  }
+
+  private _computeSuffix(schema) {
+    return this.computeSuffix
+      ? this.computeSuffix(schema)
+      : schema && schema.description
+      ? schema.description.unit_of_measurement
```
Comment (balloob) [inline src/components/form/ha-form.ts]: I wonder if this should have been `schema.description.suffix` and that we made a mistake referring to it as `unit_of_measurement`. It doesn't scale to other things.

File: src/components/form/ha-form.ts [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+  private _getValue(obj, item) {
+    if (obj) {
+      return obj[item.name];
+    }
+    return null;
+  }
+
+  private _valueChanged(ev: CustomEvent) {
+    ev.stopPropagation();
```
Comment (balloob) [inline src/components/form/ha-form.ts]: needed if none of them bubble?

File: src/panels/config/js/condition/device.tsx [hunk 15]
```diff
@@ -64,9 +64,9 @@ export default class DeviceCondition extends Component<any, any> {
         {extraFieldsData && (
           <ha-form
             data={Object.assign({}, ...extraFieldsData)}
-            onData-changed={this._extraFieldsChanged}
             schema={this.state.capabilities.extra_fields}
             computeLabel={this._extraFieldsComputeLabelCallback(hass.localize)}
+            onValue-changed={this._extraFieldsChanged}
```
Comment (balloob) [inline src/panels/config/js/condition/device.tsx]:
```suggestion
            onvalue-changed={this._extraFieldsChanged}
```

File: src/panels/config/js/condition/device.tsx [hunk 16]
```diff
@@ -98,15 +98,9 @@ export default class DeviceCondition extends Component<any, any> {
...
-    if (!ev.detail.path) {
-      return;
-    }
-    const item = ev.detail.path.replace("data.", "");
-    const value = ev.detail.value || undefined;
-
     this.props.onChange(this.props.index, {
-      ...this.props.condition,
-      [item]: value,
+      ...this.props.trigger,
```
Comment (balloob) [inline src/panels/config/js/condition/device.tsx]: should remain condition ?

**Comment by balloob** [comment]
Don't forget to register the elements in the HTML map

```ts
declare global {
  interface HTMLElementTagNameMap {
    "hui-thermostat-card": HuiThermostatCard;
  }
}
```

File: src/components/form/ha-form.ts (hunk 12 above)
Comment (bramkragten) [reply src/components/form/ha-form.ts]: Render takes care of that, `div#element` is only rendered if the schema is not an array

File: src/components/form/ha-form.ts (hunk 14 above)
Comment (bramkragten) [reply src/components/form/ha-form.ts]: Uuuhm, yeah I expected it not to be needed, but it is... 😕

**Review by balloob — APPROVED** [review]

**Pull request merged**